From ddbe65963dec1e261c23d027602b89d18d90ae63 Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Sat, 4 Feb 2023 13:58:20 -0600
Subject: [PATCH 02/68] riscv: cpu: Add skeleton for T-HEAD CPUs

This is a direct copy of the existing generic CPU code.

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 arch/riscv/cpu/thead/Makefile |  4 ++++
 arch/riscv/cpu/thead/cpu.c    | 22 ++++++++++++++++++++++
 arch/riscv/cpu/thead/dram.c   | 22 ++++++++++++++++++++++
 3 files changed, 48 insertions(+)
 create mode 100644 arch/riscv/cpu/thead/Makefile
 create mode 100644 arch/riscv/cpu/thead/cpu.c
 create mode 100644 arch/riscv/cpu/thead/dram.c

--- /dev/null
+++ b/arch/riscv/cpu/thead/Makefile
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: GPL-2.0+
+
+obj-y += cpu.o
+obj-y += dram.o
--- /dev/null
+++ b/arch/riscv/cpu/thead/cpu.c
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2018, Bin Meng <bmeng.cn@gmail.com>
+ */
+
+#include <irq_func.h>
+#include <asm/cache.h>
+
+/*
+ * cleanup_before_linux() is called just before we call linux
+ * it prepares the processor for linux
+ *
+ * we disable interrupt and caches.
+ */
+int cleanup_before_linux(void)
+{
+	disable_interrupts();
+
+	cache_flush();
+
+	return 0;
+}
--- /dev/null
+++ b/arch/riscv/cpu/thead/dram.c
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2018, Bin Meng <bmeng.cn@gmail.com>
+ */
+
+#include <common.h>
+#include <fdtdec.h>
+#include <init.h>
+#include <asm/global_data.h>
+#include <linux/sizes.h>
+
+DECLARE_GLOBAL_DATA_PTR;
+
+int dram_init(void)
+{
+	return fdtdec_setup_mem_size_base();
+}
+
+int dram_init_banksize(void)
+{
+	return fdtdec_setup_memory_banksize();
+}
